# Итерационные методы решения СЛАУ

Выполнил студент 3 курса СЦТ Лютарь Владислав

Вариант 14

In [98]:
import numpy as np

def Norm2(v):
    return sum([v_[0]*v_[0] for v_ in v])**.5

# Метод Якоби

In [99]:
def JacobiMethod(A_, b_, x0, eps):
    n = len(A_)
    A = np.zeros((n,n))
    b = np.zeros((n,1))

    kaunt = 0

    for i in range(n):
        b[i, 0] = b_[i, 0] / A_[i, i]

        for j in range(n):
            if i != j: A[i, j] = -A_[i, j] / A_[i, i]

    xbuf = np.zeros((n,1))
    xnext = x0

    while Norm2(xnext - xbuf) > eps:
        xbuf = xnext
        xnext = A @ xnext + b

        kaunt += 1
        #print(xnext)

    return xnext, kaunt


A = np.array([
    [4, 2, 1],
    [2, 4, 1],
    [1, 2, 4]
])

b = np.array([
    [1],
    [1],
    [1]
])

x = np.array([
    [-1],
    [1],
    [0]
])


x_, k = JacobiMethod(A, b, x, 0.001)

print(Norm2(x_ - np.array([[1/7], [1/7], [1/7]])), k)
print()


A = np.array([
    [6.22, 1.42, -1.72, 1.91],
    [1.42, 5.33, 1.11, -1.82],
    [-1.72, 1.11, 5.24, 1.42],
    [1.91, -1.82, 1.42, 6.55]
])

b = np.array([
    [7.53],
    [6.06],
    [8.05],
    [8.06]
])

x = np.array([
    [-1],
    [1],
    [1],
    [0]
])


x_, k = JacobiMethod(A, b, x, 0.0001)

print(x_, k)
print(Norm2(A @ x_ - b))

0.00038241800731011815 21

[[1.28739931]
 [0.69360051]
 [1.62327915]
 [0.69596615]] 38
0.00039029854287367036


# Метод верхней релаксации

In [104]:
def URM(A_, b_, x0, omega, eps):
    n = len(A_)
    
    D = np.zeros((n,n))
    U = np.zeros((n,n))
    L = np.zeros((n,n))

    kaunt = 0

    for i in range(n):

        for j in range(n):
            if i > j: 
                U[i, j] = A_[i, j]
            elif i < j:
                L[i, j] = A_[i, j]
            else:
                D[i, j] = A_[i, j]

    Lowerinv = np.linalg.inv(D + omega * L)
    bnew = omega * b_
    Upper = omega * U + (omega - 1) * D

    xbuf = np.zeros((n,1))
    xnext = x0

    while Norm2(xnext - xbuf) > eps:
        xbuf = xnext
        xnext = Lowerinv @ (bnew - Upper @ xnext)
        kaunt += 1

    return xnext, kaunt


A = np.array([
    [4, 2, 1],
    [2, 4, 1],
    [1, 2, 4]
])

b = np.array([
    [1],
    [1],
    [1]
])

x = np.array([
    [-1],
    [1],
    [0]
])

x_, k = URM(A, b, x, 1.1, 0.001)

print(Norm2(x_ - np.array([[1/7], [1/7], [1/7]])), k)
print()

A = np.array([
    [6.22, 1.42, -1.72, 1.91],
    [1.42, 5.33, 1.11, -1.82],
    [-1.72, 1.11, 5.24, 1.42],
    [1.91, -1.82, 1.42, 6.55]
])

b = np.array([
    [7.53],
    [6.06],
    [8.05],
    [8.06]
])

x = np.array([
    [-1],
    [1],
    [1],
    [0]
])

x_, k = URM(A, b, x, 1.1, 0.0001)

print(x_, k)
print(Norm2(A @ x_ - b))
print()

for i in range(100):
    x_, k = URM(A, b, x, 1 + i/100, 0.0001)
    print(f"ω: {1 + i/100}\t\t k: {k}")


0.0002504741823293033 7

[[1.28752453]
 [0.69348179]
 [1.62339034]
 [0.69586608]] 19
0.00019400572465811517

ω: 1.0		 k: 23
ω: 1.01		 k: 23
ω: 1.02		 k: 23
ω: 1.03		 k: 22
ω: 1.04		 k: 22
ω: 1.05		 k: 21
ω: 1.06		 k: 21
ω: 1.07		 k: 20
ω: 1.08		 k: 20
ω: 1.09		 k: 20
ω: 1.1		 k: 19
ω: 1.11		 k: 19
ω: 1.12		 k: 18
ω: 1.13		 k: 18
ω: 1.1400000000000001		 k: 18
ω: 1.15		 k: 17
ω: 1.16		 k: 17
ω: 1.17		 k: 16
ω: 1.18		 k: 16
ω: 1.19		 k: 16
ω: 1.2		 k: 15
ω: 1.21		 k: 15
ω: 1.22		 k: 14
ω: 1.23		 k: 14
ω: 1.24		 k: 13
ω: 1.25		 k: 13
ω: 1.26		 k: 13
ω: 1.27		 k: 12
ω: 1.28		 k: 12
ω: 1.29		 k: 11
ω: 1.3		 k: 12
ω: 1.31		 k: 12
ω: 1.32		 k: 12
ω: 1.33		 k: 12
ω: 1.34		 k: 13
ω: 1.35		 k: 13
ω: 1.3599999999999999		 k: 13
ω: 1.37		 k: 13
ω: 1.38		 k: 14
ω: 1.3900000000000001		 k: 14
ω: 1.4		 k: 14
ω: 1.41		 k: 15
ω: 1.42		 k: 15
ω: 1.43		 k: 15
ω: 1.44		 k: 16
ω: 1.45		 k: 16
ω: 1.46		 k: 17
ω: 1.47		 k: 17
ω: 1.48		 k: 18
ω: 1.49		 k: 18
ω: 1.5		 k: 18
ω: 1.51		 k: 19
ω: 1.52		 k: 19
ω: 1.53

# Резюмирование чтоли

Оба метода просты в реализации, но предпочтение вероятно стоит отдать всё таки методу верхней релаксации, т.к. при параметре ω = 1.29 он сошелся к желаемой точности за 11 итераций, в то время как метод Якоби сошелся за 21